db/log_writer.cc

```diff
@@ -56,6 +66,12 @@ IOStatus Writer::AddRecord(const Slice& slice) {
   const int header_size =
       recycle_log_files_ ? kRecyclableHeaderSize : kHeaderSize;
 
+  // If the compression type record could not be written, return the failure
+  // status here.
+  if (!add_compression_type_record_status_.ok()) {
```

Comment:
This is modifying a performance-critical path, and I think it should be checked for performance regression. I'm not expecting one, as it should be a one or two instruction check with predictable outcome, but we should be cautious, especially if a later PR is going to use a performance baseline starting after this PR.

Reply:
@pdillinger Ran the benchmarks. The numbers seem similar to me. Could you please take a look as well?

20 runs without changes - P480435381

fillseq      :       2.365 micros/op 422758 ops/sec;   46.8 MB/s
fillseq      :       2.365 micros/op 422821 ops/sec;   46.8 MB/s
fillseq      :       2.384 micros/op 419500 ops/sec;   46.4 MB/s
fillseq      :       2.340 micros/op 427433 ops/sec;   47.3 MB/s
fillseq      :       2.392 micros/op 418029 ops/sec;   46.2 MB/s
fillseq      :       2.366 micros/op 422663 ops/sec;   46.8 MB/s
fillseq      :       2.309 micros/op 433136 ops/sec;   47.9 MB/s
fillseq      :       2.355 micros/op 424697 ops/sec;   47.0 MB/s
fillseq      :       2.318 micros/op 431453 ops/sec;   47.7 MB/s
fillseq      :       2.288 micros/op 436991 ops/sec;   48.3 MB/s
fillseq      :       2.353 micros/op 424979 ops/sec;   47.0 MB/s
fillseq      :       2.358 micros/op 424174 ops/sec;   46.9 MB/s
fillseq      :       2.371 micros/op 421841 ops/sec;   46.7 MB/s
fillseq      :       2.361 micros/op 423500 ops/sec;   46.9 MB/s
fillseq      :       2.444 micros/op 409122 ops/sec;   45.3 MB/s
fillseq      :       2.350 micros/op 425470 ops/sec;   47.1 MB/s
fillseq      :       2.410 micros/op 414962 ops/sec;   45.9 MB/s
fillseq      :       2.441 micros/op 409707 ops/sec;   45.3 MB/s
fillseq      :       2.347 micros/op 426002 ops/sec;   47.1 MB/s
fillseq      :       2.486 micros/op 402202 ops/sec;   44.5 MB/s
fillseq      :       2.378 micros/op 420528 ops/sec;   46.5 MB/s

20 runs with changes - P480435504

fillseq      :       2.652 micros/op 377039 ops/sec;   41.7 MB/s
fillseq      :       2.411 micros/op 414761 ops/sec;   45.9 MB/s
fillseq      :       2.456 micros/op 407172 ops/sec;   45.0 MB/s
fillseq      :       2.466 micros/op 405458 ops/sec;   44.9 MB/s
fillseq      :       2.485 micros/op 402465 ops/sec;   44.5 MB/s
fillseq      :       2.561 micros/op 390497 ops/sec;   43.2 MB/s
fillseq      :       2.404 micros/op 416024 ops/sec;   46.0 MB/s
fillseq      :       2.407 micros/op 415491 ops/sec;   46.0 MB/s
fillseq      :       2.454 micros/op 407444 ops/sec;   45.1 MB/s
fillseq      :       2.503 micros/op 399534 ops/sec;   44.2 MB/s
fillseq      :       2.454 micros/op 407570 ops/sec;   45.1 MB/s
fillseq      :       2.600 micros/op 384638 ops/sec;   42.6 MB/s
fillseq      :       2.447 micros/op 408633 ops/sec;   45.2 MB/s
fillseq      :       2.495 micros/op 400802 ops/sec;   44.3 MB/s
fillseq      :       2.479 micros/op 403317 ops/sec;   44.6 MB/s
fillseq      :       2.546 micros/op 392845 ops/sec;   43.5 MB/s
fillseq      :       2.445 micros/op 408975 ops/sec;   45.2 MB/s
fillseq      :       2.508 micros/op 398752 ops/sec;   44.1 MB/s
fillseq      :       2.457 micros/op 406977 ops/sec;   45.0 MB/s
fillseq      :       2.503 micros/op 399529 ops/sec;   44.2 MB/s
fillseq      :       2.440 micros/op 409899 ops/sec;   45.3 MB/s
fillseq      :       2.524 micros/op 396142 ops/sec;   43.8 MB/s
fillseq      :       2.431 micros/op 411429 ops/sec;   45.5 MB/s

Reply:
Forgot to compare the averages. Doing that now.

Reply:
Did another run, 20 each.
The averages are -
Before 417291 ops/sec
After 405707 ops/sec

Not sure if that's significant. Still seeing a lot of variance in individual runs.

Before run - 
fillseq      :       2.384 micros/op 419476 ops/sec;   46.4 MB/s
fillseq      :       2.327 micros/op 429687 ops/sec;   47.5 MB/s
fillseq      :       2.492 micros/op 401243 ops/sec;   44.4 MB/s
fillseq      :       2.310 micros/op 432919 ops/sec;   47.9 MB/s
fillseq      :       2.303 micros/op 434279 ops/sec;   48.0 MB/s
fillseq      :       2.370 micros/op 421879 ops/sec;   46.7 MB/s
fillseq      :       2.324 micros/op 430227 ops/sec;   47.6 MB/s
fillseq      :       2.405 micros/op 415754 ops/sec;   46.0 MB/s
fillseq      :       2.672 micros/op 374202 ops/sec;   41.4 MB/s
fillseq      :       2.522 micros/op 396483 ops/sec;   43.9 MB/s
fillseq      :       2.534 micros/op 394696 ops/sec;   43.7 MB/s
fillseq      :       2.291 micros/op 436510 ops/sec;   48.3 MB/s
fillseq      :       2.367 micros/op 422516 ops/sec;   46.7 MB/s
fillseq      :       2.500 micros/op 400036 ops/sec;   44.3 MB/s
fillseq      :       2.367 micros/op 422452 ops/sec;   46.7 MB/s
fillseq      :       2.432 micros/op 411267 ops/sec;   45.5 MB/s
fillseq      :       2.335 micros/op 428173 ops/sec;   47.4 MB/s
fillseq      :       2.318 micros/op 431461 ops/sec;   47.7 MB/s
fillseq      :       2.372 micros/op 421660 ops/sec;   46.6 MB/s
fillseq      :       2.376 micros/op 420916 ops/sec;   46.6 MB/s

After run -
fillseq      :       2.406 micros/op 415663 ops/sec;   46.0 MB/s
fillseq      :       2.454 micros/op 407516 ops/sec;   45.1 MB/s
fillseq      :       2.487 micros/op 402063 ops/sec;   44.5 MB/s
fillseq      :       2.376 micros/op 420890 ops/sec;   46.6 MB/s
fillseq      :       2.416 micros/op 413835 ops/sec;   45.8 MB/s
fillseq      :       2.465 micros/op 405668 ops/sec;   44.9 MB/s
fillseq      :       2.395 micros/op 417504 ops/sec;   46.2 MB/s
fillseq      :       2.429 micros/op 411646 ops/sec;   45.5 MB/s
fillseq      :       2.410 micros/op 414870 ops/sec;   45.9 MB/s
fillseq      :       2.488 micros/op 401849 ops/sec;   44.5 MB/s
fillseq      :       2.434 micros/op 410790 ops/sec;   45.4 MB/s
fillseq      :       2.710 micros/op 368959 ops/sec;   40.8 MB/s
fillseq      :       2.599 micros/op 384809 ops/sec;   42.6 MB/s
fillseq      :       2.466 micros/op 405485 ops/sec;   44.9 MB/s
fillseq      :       2.426 micros/op 412143 ops/sec;   45.6 MB/s
fillseq      :       2.471 micros/op 404645 ops/sec;   44.8 MB/s
fillseq      :       2.483 micros/op 402751 ops/sec;   44.6 MB/s
fillseq      :       2.404 micros/op 416001 ops/sec;   46.0 MB/s
fillseq      :       2.654 micros/op 376793 ops/sec;   41.7 MB/s
fillseq      :       2.379 micros/op 420260 ops/sec;   46.5 MB/s

Reply:
Is the before run using db_bench from the branch, or main?

Reply:
The before is from main. This is the head -
commit c27ca2364417c96f4e56a1902cb4b800fee46f18 (HEAD -> main, origin/main, origin/HEAD)
Author: Siddhartha Roychowdhury <sidroyc@fb.com>
Date:   Wed Jan 26 13:57:30 2022 -0800

    Add option for WAL compression algorithm (#9432)

Reply:
And are you using DEBUG_LEVEL=0?

How I would like to do tests like this is use two terminals (before & after) running `db_bench -benchmarks=fillseq[-X500]` for max 500 iterations but have it give updates about the average performance, ideally including confidence intervals based on a basic analysis of variance (OK to assume normal distribution). This way, you can run it as long as needed to see a significant difference and then interrupt.

Right now, this setting only provides a summary after running through all the iterations (`combined_stats.Report(name);`).

Any interest in implementing this benchmarking enhancement?

Reply:
Also @sidroyc, I noticed you pushed a commit with a fix for wal_compression option default value in db_bench. Did your test include that fix? It seems without that fix, wal_compression would have been set to Snappy, which might have contributed some overhead to the "after" test due to the record addition.

Reply:
I haven't run it with DEBUG_LEVEL=0. Let me work on the remaining tasks and get back to this.

I have run it with the fix but since snappy is unsupported right now, I don't think it should affect the benchmark results.  I can run it a few more times to be sure.